src/frontend/src/optimizer/plan_node/utils.rs

```diff
@@ -73,23 +91,22 @@ impl TableCatalogBuilder {
             is_hidden: false,
         });
 
-        self.add_order_column(column_desc, order_type);
+        if let Some(order) = order_type {
+            self.add_order_column(i32::from(column_desc.column_id) as usize, order);
+        }
     }
 
     /// Check whether need to add a ordered column. Different from value, order desc equal pk in
     /// semantics and they are encoded as storage key.
-    fn add_order_column(&mut self, column_desc: ColumnDesc, order_type: Option<OrderType>) {
-        let index = i32::from(column_desc.column_id) as usize;
-        if let Some(order) = order_type {
-            self.pk_indices.push(index);
-            self.order_key.push(FieldOrder {
-                index,
-                direct: match order {
-                    OrderType::Ascending => Direction::Asc,
-                    OrderType::Descending => Direction::Desc,
-                },
-            });
-        }
+    pub fn add_order_column(&mut self, index: usize, order_type: OrderType) {
```

Comment:
I believe the column id or `index` should be managed and generated by the builder, instead of manually provided one... Better ideas? cc @BowenXiao1999

Reply:
Indeed, this is a little weird. Thanks for point out. `pk_indices` should be the index of column in state table instead of index in schema. 

@yuhao-su I checked the code a bit, why not check whether the index is `pk_indices` and dist_key when loop the `column_fields`?

Reply:
I think `column_fields` should include `pk_indices`?

Reply:
Got it, so it's special case for HashJoin (all input columns are just table_columns, so `pk_indices` = column index in state table). 

I suggest to add more docs on how `add_order_column` will be used publicly by HashJoin, e.g. some example above.

Reply:
Just updated some comments for  `add_column_desc_from_field_without_order_type` in the latest commit.

Reply:
Yes, but I think the most confusing part is `add_order_column` that is accept a index publicly. Also add some?